sdk/src/main/java/io/opentdf/platform/sdk/Autoconfigure.java

```diff
@@ -0,0 +1,741 @@
+                Disjunction r = (Disjunction) obj;
+                if (this.size() != r.size()) {
+                    return false;
+                }
+                for (int i = 0; i < this.size(); i++) {
+                    if (!this.get(i).equals(r.get(i))) {
+                        return false;
+                    }
+                }
+                return true;
+            }
+
+            @Override
+            public int hashCode() {
+                return super.hashCode();
+            }
+
+        }
+
+        public static boolean within(List<Disjunction> list, Disjunction e) {
+            for (Disjunction v : list) {
+                if (e.equals(v)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        public static String ruleToOperator(AttributeRuleTypeEnum e) {
+            switch (e) {
+                case ATTRIBUTE_RULE_TYPE_ENUM_ALL_OF:
+                    return "allOf";
+                case ATTRIBUTE_RULE_TYPE_ENUM_ANY_OF:
+                    return "anyOf";
+                case ATTRIBUTE_RULE_TYPE_ENUM_HIERARCHY:
+                    return "hierarchy";
+                case ATTRIBUTE_RULE_TYPE_ENUM_UNSPECIFIED:
+                    return "unspecified";
+                default:
+                    return "";
+            }
+        }
+
+    }
+
+    // Gets a list of directory of KAS grants for a list of attribute FQNs
+    public static Granter newGranterFromService(AttributesServiceFutureStub as, AttributeValueFQN... fqns) throws AutoConfigureException, InterruptedException, ExecutionException {
+        String[] fqnsStr = new String[fqns.length];
+        for (int i = 0; i < fqns.length; i++) {
+            fqnsStr[i] = fqns[i].toString();
+        }
+
+        GetAttributeValuesByFqnsRequest request = GetAttributeValuesByFqnsRequest.newBuilder()
+            .addAllFqns(Arrays.asList(fqnsStr))
+            .setWithValue(AttributeValueSelector.newBuilder().setWithKeyAccessGrants(true).build())
+            .build();
+
+        GetAttributeValuesByFqnsResponse av;
+        av = as.getAttributeValuesByFqns(request).get();
+
+        Granter grants = new Granter(Arrays.asList(fqns));
+
+        for (Map.Entry<String,GetAttributeValuesByFqnsResponse.AttributeAndValue> entry : av.getFqnAttributeValuesMap().entrySet()) {
+            String fqnstr = entry.getKey();
+            AttributeAndValue pair = entry.getValue();
+
+            AttributeValueFQN fqn;
+            try {
+                fqn = new AttributeValueFQN(fqnstr);
+            } catch (Exception e) {
+                return grants;
+            }
+
+            Attribute def = pair.getAttribute();
+            if (def != null) {
+                grants.addAllGrants(fqn, def.getGrantsList(), def);
+            }
+
+            Value v = pair.getValue();
+            if (v != null) {
+                grants.addAllGrants(fqn, v.getGrantsList(), def);
+            }
```

Comment:
We no longer add both grants from Value and Attribute. Instead, grants from Values, if they exist, overload any grants on attributes.

```suggestion
            Value v = pair.getValue();
            if (v != null && !v.getGrantsList().isEmpty()) {
                grants.addAllGrants(fqn, v.getGrantsList(), def);
            } else {
              Attribute def = pair.getAttribute();
              if (def != null) {
                  grants.addAllGrants(fqn, def.getGrantsList(), def);
              }
            }

```

Reply:
will do this as a follow on with the test from https://github.com/opentdf/platform/pull/1318/files